File: JAVA_PART/src/main/java/testing/platform/time/test/PatternMatchingRabinKarp.java
package testing.platform.time.test;

import org.apache.commons.io.IOUtils;
import testing.platform.algorithms.implementation.RabinKarp;
import testing.platform.Algorithm;

import java.io.File;
import java.io.FileInputStream;
import java.io.PrintWriter;

public class PatternMatchingRabinKarp extends Algorithm {

    String filePath = "C:\\Users\\wlato\\Desktop\\praca_inzynierska\\potop.txt";
    String text;

    public PatternMatchingRabinKarp(String fileName, int iterations) {
        super(fileName, iterations);
    }

    @Override
    public void prepareTestData() {
        try {
            text = IOUtils.toString(new FileInputStream(new File(filePath)), "UTF-8");
        } catch(Exception ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void startTimeTest() {
        prepareTestData();
        PrintWriter pw = this.prepareFileWriter();

        for (int i = 0; i < this.getIterations(); i++) {

            timeDiff = System.nanoTime();

            RabinKarp.search("Zagłoba",text, 101);
            RabinKarp.search("Kmicic",text,107);
            RabinKarp.search("Radziwiłł", text,113);
            RabinKarp.search( "Czarniecki", text,117);
            RabinKarp.search( "Kowalski", text, 119);
            RabinKarp.search( "Bóg", text, 123);
            RabinKarp.search( "Kazimierz", text, 127);
            RabinKarp.search( "Lubomirski" , text, 131);
            RabinKarp.search( "Kordecki" , text, 137);
            RabinKarp.search( "Sapieha" , text, 139);

            RabinKarp.search("Chmielnicki",text, 201);
            RabinKarp.search("Wolodyjowski",text,207);
            RabinKarp.search("Oleńka", text,213);
            RabinKarp.search( "Babinicz", text,217);
            RabinKarp.search( "Kordecki", text, 219);
            RabinKarp.search( "Zamojski", text, 223);
            RabinKarp.search( "Sakowicz", text, 227);
            RabinKarp.search( "Gustaw" , text, 231);
            RabinKarp.search( "Radziejowski" , text, 237);
            RabinKarp.search( "Zamoyski" , text, 239);

            pw.println( System.nanoTime() - timeDiff );
        }
        pw.close();
    }

    public static void main(String[] args) {
        PatternMatchingRabinKarp pmr = new PatternMatchingRabinKarp("RabinKarp", 100);
        pmr.startTimeTest();
    }
}

File: CPP_PART/testing/time/test/PatternMatchingRabinKarp.cpp
//
// Created by vm on 03.11.17.
//

#include <sstream>
#include <iomanip>
#include <vector>
#include "PatternMatchingRabinKarp.h"
#include "../../algorithms/RabinKarp.h"

PatternMatchingRabinKarp::PatternMatchingRabinKarp(const std::string &fileName, int iterations, const std::string textFilePath) : Algorithm(fileName,
                                                                                                            iterations),textFilePath(textFilePath) {}

void PatternMatchingRabinKarp::prepareTestData() {
    std::cout << textFilePath;

    std::ifstream in_file(textFilePath);
    if (!in_file) {
        std::cerr << "Could not open input file\n";
        return;
    }
    std::stringstream buffer;
    buffer << in_file.rdbuf();

    //std::cout << buffer.str();
    std::string str  = buffer.str();

    text = new char[str.length() + 1];
    strcpy(text, str.c_str());
}


void PatternMatchingRabinKarp::startTimeTest() {
    prepareTestData();
    std::fstream fs = prepareFileWriter();
    char s1[] = "Zagłoba";
    char s2[] = "Kmicic";
    char s3[] = "Radziwiłł";
    char s4[] = "Czarnecki";
    char s5[] = "Kowalski";
    char s6[] = "Wrzeszczowicz";
    char s7[] = "Kazimierz";
    char s8[] = "Lubomirski";
    char s9[] = "Kordecki";
    char s10[] = "Sapieha";

    char s11[] = "Chmielnicki";
    char s12[] = "Wołodyjowski";
    char s13[] = "Oleńka";
    char s14[] = "Babinicz";
    char s15[] = "Kordecki";
    char s16[] = "Zamoyski";
    char s17[] = "Sakowicz";
    char s18[] = "Gustaw";
    char s19[] = "Radziejowski";
    char s20[] = "Miller";


    for (int j = 0; j < iterations; ++j) {

        auto start = std::chrono::high_resolution_clock::now();
        search( s1 ,text, 101);
        search( s2 ,text, 101);
        search( s3 ,text, 101);
        search( s4 ,text, 101);
        search( s5 ,text, 101);
        search( s6 ,text, 101);
        search( s7 ,text, 101);
        search( s8 ,text, 101);
        search( s9 ,text, 101);
        search( s10 ,text, 101);
        search( s11 ,text, 101);
        search( s12 ,text, 101);
        search( s13 ,text, 101);
        search( s14 ,text, 101);
        search( s15 ,text, 101);
        search( s16 ,text, 101);
        search( s17 ,text, 101);
        search( s18 ,text, 101);
        search( s19 ,text, 101);
        search( s20 ,text, 101);


        diff = std::chrono::high_resolution_clock::now() - start;
        fs << std::setprecision(0) << std::fixed << diff.count() * 1000000000 << std::endl;
        std::cout << j << std::endl;
    }

    fs.close();

    delete[] text;
}
